politeiad/backendv2/tstorebe/plugins/dcrdata/dcrdata.go

```diff
@@ -0,0 +1,694 @@
+	return dcrdata.ScriptSig{
+		Asm: s.Asm,
+		Hex: s.Hex,
+	}
+}
+
+func convertVinFromJSONRPC(v jsonrpc.Vin) dcrdata.Vin {
+	var scriptSig *dcrdata.ScriptSig
+	if v.ScriptSig != nil {
+		s := convertScriptSigFromJSONRPC(*v.ScriptSig)
+		scriptSig = &s
+	}
+	return dcrdata.Vin{
+		Coinbase:    v.Coinbase,
+		Stakebase:   v.Stakebase,
+		Txid:        v.Txid,
+		Vout:        v.Vout,
+		Tree:        v.Tree,
+		Sequence:    v.Sequence,
+		AmountIn:    v.AmountIn,
+		BlockHeight: v.BlockHeight,
+		BlockIndex:  v.BlockIndex,
+		ScriptSig:   scriptSig,
+	}
+}
+
+func convertVinsFromV5(ins []jsonrpc.Vin) []dcrdata.Vin {
+	i := make([]dcrdata.Vin, 0, len(ins))
+	for _, v := range ins {
+		i = append(i, convertVinFromJSONRPC(v))
+	}
+	return i
+}
+
+func convertScriptPubKeyFromV5(s v5.ScriptPubKey) dcrdata.ScriptPubKey {
+	return dcrdata.ScriptPubKey{
+		Asm:       s.Asm,
+		Hex:       s.Hex,
+		ReqSigs:   s.ReqSigs,
+		Type:      s.Type,
+		Addresses: s.Addresses,
+		CommitAmt: s.CommitAmt,
+	}
+}
+
+func convertTxInputIDFromV5(t v5.TxInputID) dcrdata.TxInputID {
+	return dcrdata.TxInputID{
+		Hash:  t.Hash,
+		Index: t.Index,
+	}
+}
+
+func convertVoutFromV5(v v5.Vout) dcrdata.Vout {
+	var spend *dcrdata.TxInputID
+	if v.Spend != nil {
+		s := convertTxInputIDFromV5(*v.Spend)
+		spend = &s
+	}
+	return dcrdata.Vout{
+		Value:               v.Value,
+		N:                   v.N,
+		Version:             v.Version,
+		ScriptPubKeyDecoded: convertScriptPubKeyFromV5(v.ScriptPubKeyDecoded),
+		Spend:               spend,
+	}
+}
+
+func convertVoutsFromV5(outs []v5.Vout) []dcrdata.Vout {
+	o := make([]dcrdata.Vout, 0, len(outs))
+	for _, v := range outs {
+		o = append(o, convertVoutFromV5(v))
+	}
+	return o
+}
+
+func convertTrimmedTxFromV5(t v5.TrimmedTx) dcrdata.TrimmedTx {
+	return dcrdata.TrimmedTx{
+		TxID:     t.TxID,
+		Version:  t.Version,
+		Locktime: t.Locktime,
+		Expiry:   t.Expiry,
+		Vin:      convertVinsFromV5(t.Vin),
+		Vout:     convertVoutsFromV5(t.Vout),
+	}
+}
+
+func convertTrimmedTxsFromV5(txs []v5.TrimmedTx) []dcrdata.TrimmedTx {
+	t := make([]dcrdata.TrimmedTx, 0, len(txs))
+	for _, v := range txs {
+		t = append(t, convertTrimmedTxFromV5(v))
+	}
+	return t
+}
+
+func (p *dcrdataPlugin) cmdTxsTrimmed(payload string) (string, error) {
+	log.Tracef("cmdTxsTrimmed: %v", payload)
+
+	// Decode payload
+	var tt dcrdata.TxsTrimmed
+	err := json.Unmarshal([]byte(payload), &tt)
+	if err != nil {
+		return "", err
+	}
+
+	// Get trimmed txs
+	txs, err := p.txsTrimmedHTTP(tt.TxIDs)
+	if err != nil {
+		return "", fmt.Errorf("txsTrimmedHTTP: %v", err)
+	}
+
+	// Prepare reply
+	ttr := dcrdata.TxsTrimmedReply{
+		Txs: convertTrimmedTxsFromV5(txs),
+	}
+	reply, err := json.Marshal(ttr)
+	if err != nil {
+		return "", err
+	}
+
+	return string(reply), nil
+}
+
+func (p *dcrdataPlugin) websocketMonitor() {
+	defer func() {
+		log.Infof("Dcrdata websocket closed")
+	}()
+
+	// Setup messages channel
+	receiver := p.ws.Receive()
```

Comment:
Yuck. Can we not directly read the websocket content? These type of relay channels have the tendency to become blocked leaving the application with appearance it is working but really isn't.

Reply:
This is how the dcrdata websocket client is setup.  Its not something we have control over.

https://github.com/decred/dcrdata/blob/master/pubsub/psclient/client.go#L223

Reply:
It pings every minute to verify the connection is still active.  If the ping fails it marks the websocket as disconnected and the dcrdata plugin falls back to http queries until the websocket is succesfully reconnected.

Reply:
The dcrdata plugin also returns the connection status in certain commands like the best block command so that the caller can decide if its ok to use a potentially stale best block.  The ticketvote plugin will use stale best blocks for read commands, but will fail on writes, so that politeia doesn't completely fall over like it does now if dcrdata has intermittent connection issues.